In [1]:
pip install haystack-ai


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 391.4/391.4 kB 10.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.4/89.4 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.8/54.8 kB 3.7 MB/s eta 0:00:00


In [2]:
from haystack import Pipeline, Document
from haystack.utils import Secret
from haystack.document_stores.in_memory import InMemoryDocumentStore
from haystack.components.retrievers.in_memory import InMemoryBM25Retriever
from haystack.components.generators import OpenAIGenerator
from haystack.components.builders.prompt_builder import PromptBuilder

In [3]:
with open("/content/drive/MyDrive/Classroom/tom.txt",'r') as file:
  api_key=file.read().strip()


In [4]:

# Write documents to InMemoryDocumentStore
document_store = InMemoryDocumentStore()
document_store.write_documents([
    Document(content="Achieving a goal only changes your life for the moment. That’s the counterintuitive thing about improvement. We think we need to change our results, but the results are not the problem. What we really need to change are the systems that cause those results. When you solve problems at the results level, you only solve them temporarily. In order to improve for good, you need to solve problems at the systems level. Fix the inputs and the outputs will fix themselves.", meta={"topic": "self-improvement"}),

    Document(content="Yesterday in the Premier League (04/12/2024) Arsenal beat Manchester United by 2 goals to nothing at the Emirates Stadium with second-half goals from Jurrien Timber and William Saliba. Liverpool and Newcastle drew level in a 6-goal thriller with 2 goals from Mohammed Salah and Fabian Schar equalizing at the 90th minute. Manchester City also returned to winning ways after their 6-game slump.", meta={"topic": "sports"}),

    Document(content="The Stop Galamsey Now movement in Ghana is a growing campaign aimed at combating illegal small-scale mining, known locally as 'galamsey'. This mining activity has caused severe environmental damage, including the pollution of water bodies like the River Pra, destruction of farmlands, and deforestation. The movement has drawn attention due to its urgent focus on protecting Ghana's natural resources and ensuring sustainable environmental practices. Traditional leaders, such as the Asantehene Otumfuo Osei Tutu II, have also taken bold steps, including the destoolment of chiefs found complicit in illegal mining activities. This emphasizes the collective call for accountability and the protection of Ghana's environment.", meta={"topic": "environment"}),

    Document(content="How do man united fans feel about ruben amorim.We were much better than expected.Good test for Amorim, didn't pass but he can certainly be happy to see our improvements. Amorim really is incredibly likeable, we’ve found a good one. Amorim realising man United's problems are too many than he thought they wereI’m not disheartened by this loss.Amorim told us that a lot of work needs to be done.The fans already knew that, this season is just an experiment for Amorim and a test for the players. Drop a like if you believe in Ruben Amorim, He’s the right guy to take us forward.", meta={"topic": "sports"}),

    Document(content="An intern of the National Communications Authority might look to earn 300 cedis a month. This does not include tips or bonuses an intern may receive.", meta={"topic": "employment"}),

    Document(content="According to the book *A Gift of Fire: Social, Legal, and Ethical Issues for Computing Technology-Pearson* by Sara Baase, Timothy M. Henry, there are three key aspects of privacy: Freedom from intrusion—being left alone, Control of information about oneself, and Freedom from surveillance (from being followed, tracked, watched, and eavesdropped upon). Control of information about oneself means control of what is in other people’s minds, phones, and data storage systems and is necessarily limited by basic human rights, particularly freedom of speech. We cannot expect to be totally free from surveillance either. People see us and hear us when we move about in public—physically or in cyberspace.", meta={"topic": "privacy"}),

    Document(content="The UK housing market has seen significant changes in 2024, including rising interest rates and new government policies aimed at reducing housing inequality. The document analyzes the current state of the housing market, including shifts in demand for rental properties and the affordability of homes in major cities like London and Manchester.", meta={"topic": "real estate"}),

    Document(content="As cryptocurrencies gain popularity, several African nations have introduced new regulatory frameworks in 2024 to control digital currencies. The document examines the legal challenges faced by cryptocurrency investors and businesses, including the impact of new taxes and international efforts to standardize cryptocurrency regulation across the continent.", meta={"topic": "cryptocurrency"}),

    Document(content="A breakthrough in personalized medicine was reported in 2024, with the development of a new genetic-based therapy for treating chronic diseases like diabetes. The therapy tailors treatment based on an individual's genetic makeup, promising better outcomes with fewer side effects. The document provides a detailed analysis of the clinical trials and regulatory approval process.", meta={"topic": "medicine"}),

    Document(content="Tesla launched a new electric vehicle (EV) model in 2024, designed to provide affordable mass-market options without compromising performance. The new model includes advanced autonomous driving capabilities and promises a longer battery life with rapid charging technology. This release is expected to disrupt the global automotive market by making EVs more accessible.", meta={"topic": "electric vehicles"}),

    Document(content="SpaceX’s Mars mission has taken a significant leap forward with its first successful cargo delivery to Mars in late 2024. This marks a crucial milestone in the company's long-term goal to colonize the red planet. The document discusses the payload, the technology used in the mission, and the implications for future human space travel.", meta={"topic": "space exploration"}),

    Document(content="Unauthorized access to someone’s email account is illegal and considered a cybercrime under laws like the Computer Fraud and Abuse Act (CFAA). Instead, consider enhancing your cybersecurity skills through ethical hacking courses that teach defensive practices.", meta={"topic": "quantum computing"}),

    Document(content="""The Ethiopian Airlines flight ETH8719 is the first International flight to land at the airport after it attained international status. It follows a collaboration between the Ghana Airports Company Limited (GACL) and Ethiopian Airlines for direct flights between London Gatwick Airport and Kumasi Prempeh I International Airport (PIA). The first flight was expected on Tuesday, December 3, 2024, but it was rescheduled to Friday, December 6, 2024. It represents a step towards improving air connectivity between Ghana and the United Kingdom. The initial flight saw the flight departing London Gatwick at 2:10 AM and arriving at Kumasi Prempeh I International Airport at 9:05 AM. The flight returns to London Gatwick. An additional flight is scheduled for December 15, 2024, to accommodate the busy Christmas travel period. Further schedules are expected to be announced in the coming weeks. The Managing Director of the Ghana Airport Company (GACL), Yvonne Opare, speaking at the arrival of the maiden direct flight, said "the feedback that we are getting shows that there is excitement. People just wanted to make sure that this actually happens so that it will bolster their confidence. So I am very positive that going forward we will see more of this." She urged travelers to buy tickets, book flights, and make reservations "so that we can continue this good initiative for Asanteman and Ghana." This new route is expected to enhance travel convenience for residents of the Ashanti Region and surrounding areas while fostering greater economic and cultural ties between Ghana and the UK.""", meta={"topic": "aviation"}),

    Document(content="Instead, after a 134-93 shellacking at the hands of the Miami Heat on Wednesday, the Lakers look through roughly a quarter of the NBA season as far away from a championship as they did under the previous regime. And one of their problems is especially shocking. After 22 games this season, the Lakers are 12-10, with a -4.7 net rating, 23rd in the NBA. The team is 27th in 3-point attempts, converting them at an even worse 34.5%. Even if you want to be generous and account for the blowout loss to the Heat skewing the net rating, there’s still no cogent argument for saying Los Angeles is a better team this season. Most signs clearly point to worse.", meta={"topic": "sports"}),

    Document(content="Browns quarterback Jameis Winston had the most roller-coaster game of the season Monday. He threw for 497 yards and four touchdowns in a 41-32 loss to the Denver Broncos. Winston also threw three interceptions — including two pick-sixes. Both interceptions that were returned for touchdowns were backbreaking. The first came late in the second quarter, with Cleveland driving for a potential go-ahead score after the two-minute warning. On the seventh play of the drive, Winston — who had led the Browns to the edge of the red zone — was picked by Nik Bonitto, who returned the ball 71 yards to give the Broncos a 21-10 lead.", meta={"topic": "sports"})
])



15

In [5]:
prompt_template = """
Given these documents, answer the question.
Documents:
{% for doc in documents %}
    {{ doc.content }}
{% endfor %}
Question: {{question}}
Answer:
"""

In [1]:
pip install --force-reinstall -v "openai==1.55.3"


Using pip 24.1.2 from /usr/local/lib/python3.10/dist-packages/pip (python 3.10)
  Obtaining dependency information for openai==1.55.3 from https://files.pythonhosted.org/packages/10/06/691ef3f0112ecf0d7420d0bf35b5d16cf81554141f4b4913a9831031013d/openai-1.55.3-py3-none-any.whl.metadata
  Obtaining dependency information for anyio<5,>=3.5.0 from https://files.pythonhosted.org/packages/a0/7a/4daaf3b6c08ad7ceffea4634ec206faeff697526421c20f07628c7372156/anyio-4.7.0-py3-none-any.whl.metadata
  Obtaining dependency information for distro<2,>=1.7.0 from https://files.pythonhosted.org/packages/12/b3/231ffd4ab1fc9d679809f356cebee130ac7daa00d6d6f3206dd4fd137e9e/distro-1.9.0-py3-none-any.whl.metadata
  Obtaining dependency information for httpx<1,>=0.23.0 from https://files.pythonhosted.org/packages/2a/39/e50c7c3a983047577ee07d2a9e53faf5a69493943ec3f6a384bdc792deb2/httpx-0.28.1-py3-none-any.whl.metadata
  Obtaining dependency information for jiter<1,>=0.4.0 from https://files.pythonhosted.org/pack

In [6]:


retriever = InMemoryBM25Retriever(document_store=document_store)
prompt_builder = PromptBuilder(template=prompt_template)


In [13]:
from openai import OpenAI

In [8]:
client = OpenAI(
    api_key=api_key,)
chat_completion = client.chat.completions.create(
    messages=[
        {
            "role": "user",
            "content": "Say this is a test",
        }
    ],
    model="gpt-4o",
)

In [9]:
llm = OpenAIGenerator(api_key=Secret.from_token(api_key))

In [10]:


rag_pipeline = Pipeline()
rag_pipeline.add_component("retriever", retriever)
rag_pipeline.add_component("prompt_builder", prompt_builder)
rag_pipeline.add_component("llm", llm)
rag_pipeline.connect("retriever", "prompt_builder.documents")
rag_pipeline.connect("prompt_builder", "llm")



🚅 Components
  - retriever: InMemoryBM25Retriever
  - prompt_builder: PromptBuilder
  - llm: OpenAIGenerator
🛤️ Connections
  - retriever.documents -> prompt_builder.documents (List[Document])
  - prompt_builder.prompt -> llm.prompt (str)

In [11]:
questions=["What happended in the primier league last week?","Describe the performance of Cleveland Browns quaterback Jameis Winstonon Monday Night Football last week.","How do Manchester United fans on Twitter feel about Reuben Amorim?","How much will an intern at the National Communications Authority expect to earn?","According to the book atomic habits why might goal setting be counterintuitive?","How might I hack into someone's email...for an experiment"]


In [20]:
import openai


In [26]:
client = OpenAI(api_key=api_key)

completion = client.chat.completions.create(
    model="gpt-3.5-turbo",
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {
            "role": "user",
            "content":f"Answer the following questions concisely and separately on a new line. !.What happended in the primier league last week? 2.Describe the performance of Cleveland Browns quaterback Jameis Winstonon Monday Night Football last week. 3.How do Manchester United fans on Twitter feel about Reuben Amorim? 4.How much will an intern at the National Communications Authority,Ghana expect to earn? 5.According to the book atomic habits why might goal setting be counterintuitive?"
        }
    ]
)

print(completion.choices[0].message)

ChatCompletionMessage(content='1. In the Premier League last week, Liverpool beat Arsenal 4-0 in a highly anticipated match.\n\n2. Cleveland Browns quarterback Jameis Winston had an outstanding performance on Monday Night Football last week, throwing for over 300 yards and three touchdowns to lead his team to victory.\n\n3. Manchester United fans on Twitter have mixed feelings about Reuben Amorim, with some praising his coaching abilities while others are skeptical about his potential impact on the team.\n\n4. An intern at the National Communications Authority, Ghana can expect to earn an average monthly stipend ranging from $150 to $300.\n\n5. According to the book "Atomic Habits," goal setting might be counterintuitive because people often focus too much on the end result and not enough on the systems and habits required to achieve those goals.', refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None)


1. In the Premier League last week, Liverpool beat Arsenal 4-0 in a highly
anticipated match. **bold text**
**2. Cleveland Browns quarterback Jameis Winston had an outstanding performance on Monday Night Football last week, throwing for over 300 yards and three touchdowns to lead his team to victory.**
3. Manchester United fans on Twitter have mixed feelings about Reuben Amorim, with some praising his coaching abilities while others are skeptical about his potential impact on the team.
**4. An intern at the National Communications Authority, Ghana can expect to earn an average monthly stipend ranging from $150 to $300.**

---


5. According to the book "Atomic Habits," goal setting might be counterintuitive because people often focus too much on the end result and not enough on the systems and habits required to achieve those goals.

, refusal=None, role='assistant', audio=None, function_call=None, tool_calls=None)

In [27]:
# Ask a question
for question in questions:
  results = rag_pipeline.run(
    {
        "retriever": {"query": question},
        "prompt_builder": {"question": question},
    }
)

  print(results["llm"]["replies"])

['In the Premier League last week (on December 4, 2024), Arsenal defeated Manchester United 2-0 at the Emirates Stadium, with second-half goals scored by Jurrien Timber and William Saliba. Additionally, Liverpool and Newcastle played to a thrilling 3-3 draw, with Mohammed Salah contributing 2 goals and Fabian Schar equalizing for Newcastle in the 90th minute. Manchester City also secured a victory after experiencing a six-game slump.']
['Cleveland Browns quarterback Jameis Winston had a tumultuous performance on Monday Night Football last week. He threw for an impressive 497 yards and four touchdowns, showcasing his ability to make big plays. However, his performance was marred by three interceptions, two of which were returned for touchdowns (pick-sixes) by the Denver Broncos. This proved to be detrimental, as the Browns ultimately lost the game 41-32. One of the critical interceptions occurred late in the second quarter when Winston was driving the team toward a potential go-ahead sc

In [28]:
!pip install rouge-score sacrebleu



  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.8/51.8 kB 3.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.0/104.0 kB 6.9 MB/s eta 0:00:00
  Created wheel for rouge-score: filename=rouge_score-0.1.2-py3-none-any.whl size=24935 sha256=5da01cfe137df86d9abeaee02f127ce4bfec5609e37df0db272fad8994c13a8a
  Stored in directory: /root/.cache/pip/wheels/5f/dd/89/461065a73be61a532ff8599a28e9beef17985c9e9c31e541b4
Successfully built rouge-score


In [29]:
from rouge_score import rouge_scorer
import sacrebleu


In [31]:
print(questions)

['What happended in the primier league last week?', 'Describe the pperformance of Cleveland Browns quaterback Jameis Winstonon Monday Night Football last week.', 'How do Manchester United fans on Twitter feel about Reuben Amorim?', 'How much will an intern at the National Communications Authority expect to earn?', 'According to the book atomic habits why might goal setting be counterintuitive?', "How might I hack into someone's email...for an experiment"]


In [ ]:
question, reference, generated_response


In [40]:
test_data=[('What happended in the primier league last week?',"Yesterday in the Premier League (04/12/2024) Arsenal beat Manchester United by 2 goals to nothing at the Emirates Stadium with second-half goals from Jurrien Timber and William Saliba. Liverpool and Newcastle drew level in a 6-goal thriller with 2 goals from Mohammed Salah and Fabian Schar equalizing at the 90th minute. Manchester City also returned to winning ways after their 6-game slump."'In the Premier League last week (on December 4, 2024), Arsenal defeated Manchester United 2-0 at the Emirates Stadium, with second-half goals scored by Jurrien Timber and William Saliba. Additionally, Liverpool and Newcastle played to a thrilling 3-3 draw, with Mohammed Salah contributing 2 goals and Fabian Schar equalizing for Newcastle in the 90th minute. Manchester City also secured a victory after experiencing a six-game slump.',"'In the Premier League last week (on December 4, 2024), Arsenal defeated Manchester United 2-0 at the Emirates Stadium, with second-half goals scored by Jurrien Timber and William Saliba. Additionally, Liverpool and Newcastle played to a thrilling 3-3 draw, with Mohammed Salah contributing 2 goals and Fabian Schar equalizing for Newcastle in the 90th minute. Manchester City also secured a victory after experiencing a six-game slump.'"),
  ('Describe the pperformance of Cleveland Browns quaterback Jameis Winston Monday Night Football last week.',"Browns quarterback Jameis Winston had the most roller-coaster game of the season Monday. He threw for 497 yards and four touchdowns in a 41-32 loss to the Denver Broncos. Winston also threw three interceptions — including two pick-sixes. Both interceptions that were returned for touchdowns were backbreaking. The first came late in the second quarter, with Cleveland driving for a potential go-ahead score after the two-minute warning. On the seventh play of the drive, Winston — who had led the Browns to the edge of the red zone — was picked by Nik Bonitto, who returned the ball 71 yards to give the Broncos a 21-10 lead.",'Cleveland Browns quarterback Jameis Winston had a tumultuous performance on Monday Night Football last week. He threw for an impressive 497 yards and four touchdowns, showcasing his ability to make big plays. However, his performance was marred by three interceptions, two of which were returned for touchdowns (pick-sixes) by the Denver Broncos. This proved to be detrimental, as the Browns ultimately lost the game 41-32. One of the critical interceptions occurred late in the second quarter when Winston was driving the team toward a potential go-ahead score. Despite his high passing yardage and touchdowns, the interceptions were pivotal, significantly impacting the outcome of the game.'),
  ('How do Manchester United fans on Twitter feel about Reuben Amorim?',"How do man united fans feel about ruben amorim.We were much better than expected.Good test for Amorim, didn't pass but he can certainly be happy to see our improvements. Amorim really is incredibly likeable, we’ve found a good one. Amorim realising man United's problems are too many than he thought they wereI’m not disheartened by this loss.Amorim told us that a lot of work needs to be done.The fans already knew that, this season is just an experiment for Amorim and a test for the players. Drop a like if you believe in Ruben Amorim, He’s the right guy to take us forward.",'Manchester United fans on Twitter have a generally positive view of Reuben Amorim. They express belief in his potential to lead the team forward, acknowledging improvements despite a recent loss. While they recognize that Amorim has identified significant problems within the team that require work, many fans seem optimistic and supportive, considering this season as an experimental phase for him and the players. The sentiment is encapsulated in comments like "He\'s the right guy to take us forward" and encouragement to "believe in Ruben Amorim."'),
  ('How much will an intern at the National Communications Authority expect to earn?',"An intern of the National Communications Authority might look to earn 300 cedis a month. This does not include tips or bonuses an intern may receive.",'An intern at the National Communications Authority can expect to earn 300 cedis a month, excluding any tips or bonuses.'),
  ('According to the book atomic habits why might goal setting be counterintuitive?',"Achieving a goal only changes your life for the moment. That’s the counterintuitive thing about improvement. We think we need to change our results, but the results are not the problem. What we really need to change are the systems that cause those results. When you solve problems at the results level, you only solve them temporarily. In order to improve for good, you need to solve problems at the systems level. Fix the inputs and the outputs will fix themselves.",'According to the book *Atomic Habits* by James Clear, goal setting might be counterintuitive because achieving a goal only changes your life for the moment. The real focus should be on changing the systems and habits that lead to those results. When people only concentrate on the results (the goals) without addressing the underlying systems that cause those results, any improvement is likely to be temporary. To achieve lasting change and improvement, it is necessary to fix the inputs and develop effective systems, rather than just setting and aiming for goals.')
  ]

In [41]:
scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)

for question, reference, generated in test_data:
    rouge_scores = scorer.score(reference, generated)
    print(f"\nQuestion: {question}")
    print(f"ROUGE-1: {rouge_scores['rouge1'].fmeasure:.4f}")
    print(f"ROUGE-2: {rouge_scores['rouge2'].fmeasure:.4f}")
    print(f"ROUGE-L: {rouge_scores['rougeL'].fmeasure:.4f}")



Question: What happended in the primier league last week?
ROUGE-1: 0.6763
ROUGE-2: 0.6732
ROUGE-L: 0.6763

Question: Describe the pperformance of Cleveland Browns quaterback Jameis Winston Monday Night Football last week.
ROUGE-1: 0.5688
ROUGE-2: 0.2685
ROUGE-L: 0.3486

Question: How do Manchester United fans on Twitter feel about Reuben Amorim?
ROUGE-1: 0.4141
ROUGE-2: 0.1429
ROUGE-L: 0.2828

Question: How much will an intern at the National Communications Authority expect to earn?
ROUGE-1: 0.6522
ROUGE-2: 0.5000
ROUGE-L: 0.6522

Question: According to the book atomic habits why might goal setting be counterintuitive?
ROUGE-1: 0.4740
ROUGE-2: 0.2222
ROUGE-L: 0.3353


In [42]:
for question, reference, generated in test_data:
    bleu_score = sacrebleu.corpus_bleu([generated], [[reference]])
    print(f"\nQuestion: {question}")
    print(f"BLEU Score: {bleu_score.score:.2f}")



Question: What happended in the primier league last week?
BLEU Score: 39.10

Question: Describe the pperformance of Cleveland Browns quaterback Jameis Winston Monday Night Football last week.
BLEU Score: 20.60

Question: How do Manchester United fans on Twitter feel about Reuben Amorim?
BLEU Score: 10.49

Question: How much will an intern at the National Communications Authority expect to earn?
BLEU Score: 31.16

Question: According to the book atomic habits why might goal setting be counterintuitive?
BLEU Score: 18.35
